### PR TITLE
feat(rust): add rendezvous service for nat hole punching

### PR DESCRIPTION
**Pull request by neil2468**
## Proposed changes

This is part of the prototype for UDP NAT Hole Punching (#3507).

- Add a 'Rendezvous service' to the UDP transport.
- Add tests to exercise new code.

Thanks!

## Checks

- [x] All commits in this Pull Request are [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) and Verified by Github.
- [x] All commits in this Pull Request follow the Ockam [commit message convention](https://github.com/build-trust/.github/blob/main/CONTRIBUTING.md#commit-messages).
- [x] I accept the Ockam Community [Code of Conduct](https://github.com/build-trust/.github/blob/main/CODE_OF_CONDUCT.md).
- [x] I have accepted the Ockam [Contributor License Agreement](https://github.com/build-trust/ockam-contributors/blob/main/CLA.md) by adding my Git/Github details in a row at the end of the [CONTRIBUTORS.csv](https://github.com/build-trust/ockam-contributors/blob/main/CONTRIBUTORS.csv) file in a separate pull request to the [build-trust/ockam-contributors](https://github.com/build-trust/ockam-contributors) repository. The easiest way to do this is to [edit the CONTRIBUTORS.csv](https://github.com/build-trust/ockam-contributors/edit/main/CONTRIBUTORS.csv) file in the github web UI and create a PR, this will mark the commit as verified.

### CODE REVIEW

File: implementations/rust/ockam/ockam_transport_udp/src/rendezvous_service/messages.rs [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+        node_name: String,
+    },
+}
+
+/// Response type for UDP Hole Punching Rendezvous service
+#[derive(Serialize, Deserialize, Debug, Message)]
+pub enum RendezvousResponse {
+    Update(Result<()>),
+    Query(Result<SocketAddr>),
+}
```
Comment (mrinalwadhwa) [inline implementations/rust/ockam/ockam_transport_udp/src/rendezvous_service/messages.rs:24]: This looks like a tiny protocol.

Leaving this comment just to note that we don't want the message serialization to be coupled with Rust's default binary encoding, so we should turn this into CBOR messages.

**Comment by mrinalwadhwa** [comment]
@neil2468 This looks awesome. Did you happen to test it across NATs somehow?

**Comment by neil2468** [comment]
@mrinalwadhwa Thanks. I've not yet tried setting up an evironment for NAT testing, but now that you mention it maybe I should do that next. 

This PR does not implement hole punching yet, it's only adds the Rendezvous service which will maintain a table of the public IPs of remote nodes.

**Review by davide-baldo — DISMISSED** [review]
Looks good!

File: implementations/rust/ockam/ockam_transport_udp/src/rendezvous_service/messages.rs (hunk 1 above)
Comment (neil2468) [reply implementations/rust/ockam/ockam_transport_udp/src/rendezvous_service/messages.rs:24]: @mrinalwadhwa Are you saying that this should change to use the `minicbor` encoding and `ockam_core::api::Request` like `CredentialExchangeWorker` does?

So I can learn, can I ask why would we like this to use CBOR while other messages (e.g. `PortalMessage` used by `TcpOutletListenWorker`) use Rust's default encoding?

Thanks!

**Review by davide-baldo — APPROVED** [review]

File: implementations/rust/ockam/ockam_transport_udp/src/rendezvous_service/messages.rs (hunk 1 above)
Comment (etorreborre) [reply implementations/rust/ockam/ockam_transport_udp/src/rendezvous_service/messages.rs:24]: @neil2468 I think that we are supposed to move everything to minicbor now. I am going to raise this issue.

Comment (etorreborre) [reply implementations/rust/ockam/ockam_transport_udp/src/rendezvous_service/messages.rs:24]: @neil2468 I actually still need to come back to you with a better answer. The question of migrating ockam messages from BARE to CBOR [is still open](https://github.com/build-trust/ockam/issues/2496). I think that what @mrinalwadhwa meant was that for workers implementing a request/response protocol, they should follow the pattern established by workers like the `CredentialExchangeWorker` and use the `Request` API present in `ockam_core`. I will come back with a confirmation.

**Pull request merged**